From 89644ad92fee38706430c1adb1e29b9756193b4f Mon Sep 17 00:00:00 2001
From: Allan Sandfeld Jensen <allan.jensen@qt.io>
Date: Thu, 17 Oct 2019 11:53:05 +0200
Subject: [PATCH] [Backport] Security issue 981459
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Don't assume HTMLFrameOwnerElement::GetLayoutObject is LayoutEmbeddedContent

The comment above HTMLFrameOwnerElement::GetLayoutEmbeddedContent()
says:
  // Most subclasses use LayoutEmbeddedContent (either LayoutEmbeddedObject or
  // LayoutIFrame) except for HTMLObjectElement and HTMLEmbedElement which may
  // return any LayoutObject when using fallback content.
  LayoutEmbeddedContent* GetLayoutEmbeddedContent() const;

No new test because I couldn't reproduce the bug locally, but I believe this
will fix the clusterfuzz bug.

Bug: 981459
Change-Id: I3ecf8022111dc25a2e862c0311ffa56467d18c2e
Reviewed-by: Philip Rogers <pdr@chromium.org>
Commit-Queue: Xianzhu Wang <wangxianzhu@chromium.org>
Cr-Commit-Position: refs/heads/master@{#675712}
Reviewed-by: Jüri Valdmann <juri.valdmann@qt.io>
Reviewed-by: Michael Brüning <michael.bruning@qt.io>
---
 .../blink/renderer/core/html/html_frame_owner_element.cc    | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/3rdparty/chromium/third_party/blink/renderer/core/html/html_frame_owner_element.cc b/src/3rdparty/chromium/third_party/blink/renderer/core/html/html_frame_owner_element.cc
index 645cbeef2cb..a2ab5eb1a36 100644
--- a/src/3rdparty/chromium/third_party/blink/renderer/core/html/html_frame_owner_element.cc
+++ b/src/3rdparty/chromium/third_party/blink/renderer/core/html/html_frame_owner_element.cc
@@ -296,8 +296,7 @@ void HTMLFrameOwnerElement::SetEmbeddedContentView(
 
   GetDocument().GetRootScrollerController().DidUpdateIFrameFrameView(*this);
 
-  LayoutEmbeddedContent* layout_embedded_content =
-      ToLayoutEmbeddedContent(GetLayoutObject());
+  LayoutEmbeddedContent* layout_embedded_content = GetLayoutEmbeddedContent();
   if (!layout_embedded_content)
     return;
 
@@ -324,8 +323,7 @@ EmbeddedContentView* HTMLFrameOwnerElement::ReleaseEmbeddedContentView() {
     return nullptr;
   if (embedded_content_view_->IsAttached())
     embedded_content_view_->DetachFromLayout();
-  LayoutEmbeddedContent* layout_embedded_content =
-      ToLayoutEmbeddedContent(GetLayoutObject());
+  LayoutEmbeddedContent* layout_embedded_content = GetLayoutEmbeddedContent();
   if (layout_embedded_content) {
     if (AXObjectCache* cache = GetDocument().ExistingAXObjectCache())
       cache->ChildrenChanged(layout_embedded_content);
-- 
2.24.0

